chapters/classes.tex

```diff
@@ -276,18 +276,71 @@ \subsubsection{Prefix Rules}\label{prefix-rules}
-Thus it is not possible to introduce equations for constants by cyclic dependencies.
+For a constant or parameter \lstinline!v! with declaration equation, the unexpanded expression of the declaration equation in the flattended model must not depend on \lstinline!v! itself, neither directly nor indirectly via other variables' declaration equations.
+To satisfy this condition, dependencies shall be removed as needed by applying simplifications based on values of constants (except with \lstinline!Evaluate = false!) and all other \willbedefined{structural parameters} (\cref{component-variability-prefixes-discrete-parameter-constant}) that don't depend on \lstinline!v!.
 
-Structural parameters, on the other hand, cannot have cyclic dependencies by definition (see \cref{component-variability-prefixes-discrete-parameter-constant}).
-Since variability rules do not allow constants to depend on structural parameters, this means that constants and structural parameters share the property that their declaration equations can be solved one at a time with the natural causality.
+That the value of a structural parameter is used for these simplifications does not mean that it has to be determined during translation, but if \lstinline!v! is found to be a structural parameter, then a Modelica tool will be able to break all cycles involving \lstinline!v! by making some (possibly none or all) of the other structural parameters determined during translation.
+Hence, evaluation of a constant or structural parameter can never require solving systems of equations; they can always be sorted so that they can be solved one at a time with the natural causality (i.e., the declaration equation is used to determine the value of the component to which it belongs).
 
 \begin{example}
+Direct and indirect cyclic dependency:
 \begin{lstlisting}[language=modelica]
-constant Real p = 2 * q;
-constant Real q = sin(p); // Illegal since p = 2 * q, q = sin(p) are cyclical
+/* All of the following are illegal: */
+parameter Real r = 2 * sin(r); // Depends directly on r.
+parameter Real p = 2 * q;      // Indirect dependency on p via q = sin(p).
+parameter Real q = sin(p);     // Indirect dependency on q via p = 2 * q.
+\end{lstlisting}
+\end{example}
+
+\begin{example}
+While declaration equations must not be cyclical, the use of initial equations can still introduce valid cyclic dependencies between parameters:
+\begin{lstlisting}[language=modelica]
+  parameter Real p = 2 * q; // This is the only declaration equation.
+  parameter Real q(fixed = false);
+initial equation
+  q = sin(p); // OK, not a declaration equation.
+\end{lstlisting}
+\end{example}
+
+\begin{example}
+Breaking cyclic dependency.
+\begin{lstlisting}[language=modelica]
+model ABCD
+  parameter Real A[n, n];
+  parameter Integer n = size(A, 1);
+end ABCD;
+
+final ABCD a;
+// Illegal cyclic dependency between size(a.A, 1) and a.n.
+
+ABCD b(redeclare Real A[2, 2] = [1, 2; 3, 4]);
```

Comment:
I know the example is from the current specification, but I just realized that this example is actually illegal because A is not replaceable.

Reply:
That's actually allowed according to 7.3.3 as far as I know, replaceable is only necessary when changing the type of the element (and changing array dimensions apparently does not count as changing the type).

Reply:
Aren't dimensions part of the type in that sense? See https://specification.modelica.org/maint/3.4/Ch10.html#array-declarations

> The number of dimensions and the dimensions sizes are part of the type, and shall be checked for example at redeclarations

Reply:
There is non-normative text in 7.3.3 saying:
>[*Redeclaration with the same type can be used to restrict variability and/or change array dimensions.*]

And according to 7.3, if no variability prefix is given the variability does not change either. I thought it would become a variable.